src/pal/pal_apple.h

```diff
@@ -1,92 +1,218 @@
-     *
-     * Currently, these are identical to the generic BSD PAL.  This field is
-     * declared explicitly to remind anyone who modifies this class that they
-     * should add any required features.
      */
-    static constexpr uint64_t pal_features = PALBSD::pal_features;
+    static constexpr uint64_t pal_features =
+      AlignedAllocation | LazyCommit | Entropy;
 
-    /**
-     * Anonymous page tag ID.
+    static constexpr size_t page_size = Aal::aal_name == ARM ? 0x4000 : 0x1000;
+
+    static constexpr size_t minimum_alloc_size = page_size;
+
+    /*
+     * Memory Tag
+     *
+     * A memory tag is an 8-bit value that denotes auxillary "type information"
+     * of a vm region. This tag can be used for marking memory for profiling and
+     * debugging, or instructing the kernel to perform tag-specific behavior.
+     * (E.g. VM_MEMORY_MALLOC{_*} is reused by default, unless it is no longer
+     * in its "original state". See `vm_map_entry_is_reusable` in
+     * `osfmk/vm/vm_map.c` for more details of this behavior.)
+     *
+     * Memory tags are encoded using `VM_MAKE_TAG(tag_value)`, and can be passed
+     * to the kernel by either `mmap` or `mach_vm_map`:
+     * 1. `fd` argument of `mmap`.
+     * 2. `flags` argument of `mach_vm_map`.
+     *
+     * There are currently 4 categories of memory tags:
+     *
+     * 1. Reserved: [0, 39]. Typically used for Apple libraries and services.
+     * Use may trigger undocumented kernel-based behavior.
+     *
+     * 2. Defined "placeholders": [39, 98]. Typically used for Apple libraries
+     * and services.
+     *
+     * 3. Undefined "placeholders": [99, 239]. Unallocated by Apple. Typically
+     * used for libraries. (E.g. LLVM sanitizers use 99.)
+     *
+     * 4. Application specific: [240, 255]
+     *
+     * See <mach/vm_statistics.h> for more details about memory tags and their
+     * uses.
+     *
+     * In the future, we may switch our default memory tag from "category 4" to
+     * "category 3", thereby affording us a "well-known" memory tag that can be
+     * easily identified in tools such as vmmap(1) or Instruments.
      *
-     * Darwin platform allows to gives an ID to anonymous pages via
-     * the VM_MAKE_TAG's macro, from 240 up to 255 are guaranteed
-     * to be free of usage, however eventually a lower could be taken
-     * (e.g. LLVM sanitizers has 99) so we can monitor their states
-     * via vmmap for instance. This value is provided to `mmap` as the file
-     * descriptor for the mapping.
      */
-    static constexpr int anonymous_memory_fd = VM_MAKE_TAG(PALAnonID);
 
-    /**
-     * Note: The root's implementation works fine on Intel
-     * however mprotect/PROT_NONE fails on ARM
-     * especially since the 11.2 release (seems known issue
-     * spotted in various projects; might be a temporary fix).
-     */
-    template<bool page_aligned = false>
-    static void zero(void* p, size_t size) noexcept
```

Comment:
We seem to have lost this function.  @mjp41, is it no longer needed?  If so, we should update the porting document to remove it and remove it from all of the other PALs.

Reply:
It is coming from the PALBSD one, but I think we should keep this version.  Good catch

Reply:
@devnexen, who opened #278, tested this branch on both Intel and M1 machines and found no regression.

About a week ago, I crawled XNU on sourcegraph in an attempt to diagnose root cause(s) for the M1 bugs, as well as reconcile the difference between the x86_64 and ARM codepaths of PALApple:

1. The Apple silicon/ARM bugs were due to PALPOSIX::page_size being defined as Aal::smallest_page_size, which is 0x1000 for ARM. It needs to be 16KiB for Apple user-space on ARM. There's quite a bit of "creative" engineering going on in XNU as of macOS 11.3 . On ARM, the kernel uses 4KiB pages internally, but requires userspace to use 16KiB pages, unless the process is "exotic" (i.e. rosetta).
2. MADV_FREE does not behave as expected. I'm not going to write the entire call tree, but the pages are never marked as "reusable" by the kernel and this can be observed through profiling. I tested MADV_FREE and MADV_FREE_REUSABLE. It is significant. At least ~75% to ~90% less dirty memory is used by func-malloc-16 (observed on x86_64). Apple's own malloc implementation as well as many ports for Apple Operating Systems use MADV_FREE_REUS{E, ABLE} instead of MADV_FREE. See https://bugs.chromium.org/p/chromium/issues/detail?id=713892   https://opensource.apple.com/source/libmalloc/libmalloc-53.1.1/src/magazine_malloc.c.auto.html


---

The reason why `notify_using` and `zero` look like duplicate functionality is b/c of optimization. Also related to another one of @davidchisnall comments below, **when `PALPOSIX::zero` calls `mmap`, it can fail and call `bzero` on the same pages that mmap failed to overwrite**.

In pseudocode, for brevity.

master:
```
PALPOSIX::zero<page_aligned>:
    if constexpr (zero_mem == YesZero)
        if page_aligned || is_aligned_block<OS::page_size>(p, size)
            mmap
            if r != MAP_FAILED
                return;
    bzero

PALPOSIX::notify_using:
    madvise(MADV_FREE)
    mprotect(PROT_READ|PROT_WRITE)
    zero<true>
```

pal-apple:
```
PALApple::notify_using:
    if constexpr (zero_mem == YesZero)
        mmap
        if r != MAP_FAILED
            return;
    mprotect(PROT_READ|PROT_WRITE)
    madvise(MADV_FREE_REUSE)
    bzero
```

On macOS (and iOS) we are free to move the mmap before mprotect, thereby eliminating the other 2 syscalls. Otherwise it would be `mprotect`, `mmap`, which is wasteful.

Reply:
Thanks, that makes a lot of sense.  It would be great to have a comment explaining this.